### PR TITLE
feat: add document viewer widget

### LINKED ISSUE #3358: [Feature] PDF viewer widget
## Summary
Users need a widget to display and print PDFs

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> document view widget add 

Fixes #3358 

## Type of change

- Bugfix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: feature/new-document-viewer 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.09 **(-0.01)** | 37.05 **(0.02)** | 33.86 **(0)** | 55.58 **(0)**
 :red_circle: | app/client/src/selectors/commentsSelectors.ts | 83.61 **(-1.64)** | 61.76 **(-2.95)** | 73.33 **(0)** | 88.24 **(-2.35)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.04 **(-0.22)** | 40.42 **(-0.83)** | 34.48 **(0)** | 55.96 **(-0.26)**
 :sparkles: :new: | **app/client/src/widgets/DocumentViewerWidget/constants.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/DocumentViewerWidget/index.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/DocumentViewerWidget/component/index.tsx** | **40.37** | **45.65** | **16.67** | **42.11**
 :sparkles: :new: | **app/client/src/widgets/DocumentViewerWidget/widget/index.tsx** | **96.77** | **100** | **80** | **96.67**</details>

### CODE REVIEW

**Review by somangshu — CHANGES_REQUESTED** [review]
@techbhavin there is a build error, Please fix

**Review by somangshu — APPROVED** [review]
Approving to run the tests once

**Comment by somangshu** [comment]
Checked out the deploy preview and the changes as per the discussion seem to reflect

File: app/client/src/widgets/DocumentViewerWidget/component/index.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      validExtension: validExtension,
+      extension: ext,
+    };
+  } else {
+    // might be preview url
+    return { hasExtension: false };
+  }
+};
+
+const getSettings = (docUrl: string) => {
```
Comment (jsartisan) [inline app/client/src/widgets/DocumentViewerWidget/component/index.tsx:39]: hey, paul, possible to add comments on these functions? for e.g - from the function name, I am not sure what settings we are talking about. I would suggest adding a brief note about what function is doing.

**Review by jsartisan — DISMISSED** [review]

**Comment by jsartisan** [comment]
LGTM! just left a small comment.

**Comment by jsartisan** [comment]
One more thing, the buttons in the PDF view looks odd one out. I guess we can't really do anything about that since we are using third-library to render the pdf viewer.

@somangshu What do you think? is that okay right?

File: app/client/src/widgets/DocumentViewerWidget/component/index.tsx (hunk 1 above)
Comment (somangshu) [reply app/client/src/widgets/DocumentViewerWidget/component/index.tsx:39]: Another point, we should write jest test for this

**Comment by somangshu** [comment]
@jsartisan the button is part of the browser default, we are using iframe from the PDF, you are right that its not in our control, We can leave it as it is.

**Comment by somangshu** [comment]
@techbhavin is the documentation done for this widget

**Comment by techbhavin** [comment]
Document is update [Here](https://github.com/appsmithorg/appsmith-docs/pull/206)

cc: @somangshu

File: app/client/src/widgets/DocumentViewerWidget/component/index.tsx [hunk 2]
```diff
@@ -36,7 +36,18 @@ const checkUrlExtension = (docUrl: string) => {
...
+export const getDocViewerConfigs = (docUrl: string) => {
+  /**
+   * From user provided document url decide on which viewer to use
+   * handle different cases for url hosts
+   * Case 1: Dropbox
+   *  if user exported download url than change that to raw url
+   * Case 2: Onedrive
+   *  change url to embeded url to show in viewer
+   * Case 3:
+   *  All other urls will be either Google Preview or urls with file extentions ( e.x. from aws s3 )
```
Comment (somangshu) [inline app/client/src/widgets/DocumentViewerWidget/component/index.tsx:48]: Here we could tell the dev about the file extensions we allow as well. Not very important

**Review by somangshu — APPROVED** [review]
LGTM, Approving to run tests

**Review by jsartisan — DISMISSED** [review]

**Comment by somangshu** [comment]
@techbhavin as mentioned, can you check if We could use the file-picker to upload the PDF and use the base64 format. Basically need the base64 URL to work here

**Comment by techbhavin** [comment]
@somangshu  @Nikhil-Nandagopal  

Current Implementation is supported base64  URL 
image

**Comment by Nikhil-Nandagopal** [comment]
@techbhavin how will a user know that URL supports base64? Please change the field name and placeholder like in the image widget

**Comment by somangshu** [comment]
@vibhandikyash merge the latest changes here and see if the icon is still valid, We will get in touch with design again otherwise

@vivekverma2312 this widget is done, Can we test this out

**Comment by YogeshJayaseelan** [comment]
@techbhavin Tested this PR and please find the observations below,

Issue 1 : The Icon of Document viewer in entity explorer is not visible clearly 

Screenshot 2021-10-12 at 10 04 20 AM

Issue 2 : The Expected Structure should be URL instead of String in the - Document link field 

Screenshot 2021-10-12 at 10 06 37 AM

Issue 3 : Passing some of the sample pdf url does not load

Url that doesn't load:

http://www.africau.edu/images/default/sample.pdf
http://www.pdf995.com/samples/pdf.pdf

Screenshot 2021-10-12 at 10 20 30 AM

Issue 4 : PDF uploaded via file picker and binded with document viewer does not work

https://www.loom.com/share/f76fdaaf19db418ab5dc149f7f6bf32d

Issue 5: Some of the pdf urls are 'refused to connect' not sure if this is an expected behaviour 

https://github.com/tpn/pdfs/blob/master/A%20Brief%20Tutorial%20on%20Database%20Queries%2C%20Data%20Mining%2C%20and%20OLAP%20(hamel-197-manuscript-final).pdf

Screenshot 2021-10-12 at 10 31 41 AM

Suggestion 1 : To have Widget wise download toggle - Though we have page wise download option we can have download toggle from widget level

Suggestion : To have search toggle - so that user will be able to search content from the PDF

Suggestion 3 : To have 'Pop up and Text notes' so that whenever user select some text - with the help of pop up user can highlight particular content on the PDF and add notes

Suggestion 4 :

**Comment by somangshu** [comment]
@YogeshJayaseelan this is document viewer which mean its can also load docx, xlsx, ppt. Do check them out

**Comment by somangshu** [comment]
@techbhavin looks like there is a related test case also failing

**Comment by YogeshJayaseelan** [comment]
@techbhavin - Tested this PR and found issue 6 is still not working 

New issue:

Issue 7 - By editing the document url to any single character - Appmsith page is loaded inside Document viewer widget 

https://www.loom.com/share/575267f7d1744409a1cac7d602c1d9f8

**Comment by techbhavin** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1434479479

File: app/client/src/widgets/DocumentViewerWidget/component/DocViewer.tsx [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+const StyledViewer = styled.div`
+  width: 100%;
+  height: 100%;
+  background: #fff;
+  overflow: auto;
+`;
+export default function DocViewer(props: { blob?: Blob }) {
+  const [state, setState] = useState({ isLoading: false, isError: false });
+  const [htmlContent, setHtmlContent] = useState("");
+  useEffect(() => {
```
Comment (jsartisan) [inline app/client/src/widgets/DocumentViewerWidget/component/DocViewer.tsx:15]: can we add some comments above useffect what we are doing in thr. just some brief info would be fine.

**Review by jsartisan — APPROVED** [review]

**Comment by techbhavin** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1444044157

**Comment by techbhavin** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1452231048

**Comment by shwetha-ramesh** [comment]
@techbhavin retested all the issues again. Here are the few things noticed.
Issue6 :  CSV, PPT, files are uploaded via file picker and binded with Document viewer - does not load the files
Currently only pdf, doc and xlxs files are supported. 
For PPT
https://loom.com/share/44e5cfb0924e46bf981c0e0eb67ddb15

For CSV
https://www.loom.com/share/c856c6b1d53747d2a180677b87864bfb

Issue 7: Still exists
https://www.loom.com/share/fe0ab983afee4ae985b8c68085d1a21a

**New Issue**
Issue 8: When visible is Enabled it becomes greyed out again disable it and the document viewer doesnt appear bold, only when the user clicks on the canvas the widget appears (refer video)

https://loom.com/share/4c05c74aa7a84a84bf11473796d2a5b6

**Comment by shwetha-ramesh** [comment]
> @YogeshJayaseelan this is document viewer which mean its can also load docx, xlsx, ppt. Do check them out

@somangshu Can you please re-confirm all the supported files. Currently only pdf, doc and xlxs files are supported. CSV and PPT are not supported.

**Comment by somangshu** [comment]
@shwetha-ramesh they should still be supported with the URL format, Maybe base64 can be a problem.

@techbhavin for issue 7 listed above, I think we should handle it, Could the widget can show that its an invalid URL?

**Comment by somangshu** [comment]
@techbhavin could you check the failing tests

**Review by somangshu — APPROVED** [review]
Failing test resolve, @jsartisan please check the code change

**Comment by somangshu** [comment]
@wmdev0808 related test is failing, [check this out](https://dashboard.cypress.io/projects/eyxvp8/runs/12061/test-results/30783d6a-1a84-4bc1-b038-f93471af73da?statuses=%5B%7B%22value%22%3A%22FAILED%22%2C%22label%22%3A%22FAILED%22%7D%5D)

**Comment by jsartisan** [comment]
@NandanAnantharamu Asked nandan about the change. If the tests are working fine, then it's okay.

**Review by jsartisan — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin Please find the testing status 

QA Comment
Issue 1 - Fixed
Issue 2 -Not in Scope for now
Issue 3 - Agreed that http is not allowed and only https are considered
Issue 4 - Verified with the right syntax Filepicker1.files[0].data and working as expected
Issue 5 - Invalid URL

Issue 6 - **Still CSV and PPT are not supported** - **Can you help us with CSV and PPT url link?**

Issue 7 - **Issue still exists** 

7.A) - Removing the // next to https or removing https still loads the appmsith site 

https://www.loom.com/share/1d5f4cb185ff424db57c7b8977781cc9

Issue 8 - **Fixed**

**Comment by YogeshJayaseelan** [comment]
@techbhavin Please find the updated status

QA Comment
Issue 1 - Fixed
Issue 2 -Not in Scope for now
Issue 3 - Agreed that http is not allowed and only https are considered
Issue 4 - Verified with the right syntax Filepicker1.files[0].data and working as expected
Issue 5 - Invalid URL
Issue 6 - **PPT is supported with the following url you have provided** 

https://roteemealplancover.s3.ap-south-1.amazonaws.com/sample/2038800838373569079sample.ppt

**But ppt file uploaded via file-picker and binded with document viewer is not supporting 

CSV - Not is scope now** 

Issue 7 - Issue still exists

7.A) - Removing the // next to https or removing https still loads the appmsith site
https://loom.com/share/1d5f4cb185ff424db57c7b8977781cc9

Issue 8 - Fixed

**Review by jsartisan — APPROVED** [review]

**Review by jsartisan — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin Verified the fix of rissue 7 and please find the status below,

QA Comment
Issue 1 - Fixed
Issue 2 -Not in Scope for now
Issue 3 - Agreed that http is not allowed and only https are considered
Issue 4 - Verified with the right syntax Filepicker1.files[0].data and working as expected
Issue 5 - Invalid URL
Issue 6 - **PPT is supported with the following url you have provided**

https://roteemealplancover.s3.ap-south-1.amazonaws.com/sample/2038800838373569079sample.ppt

**But ppt file uploaded via file-picker and binded with document viewer is not supporting**

**CSV - Not is scope now**

Issue 7 - Fixed
Issue 8 - Fixed

**Review by jsartisan — APPROVED** [review]

**Review by Frederyuyyhj — APPROVED** [review]
app/client/cypress/fixtures/DocumentViewerDsl.json

**Comment by Frederyuyyhj** [comment]
> app/client/cypress/fixtures/DocumentViewerDsl.json

**Comment by wealthz01** [comment]
You guys keep mute when j needs you

**Pull request merged**